circuit_passes/src/passes/loop_unroll/body_extractor.rs

```diff
@@ -74,115 +84,184 @@ impl SubcmpSignalCompare {
+        ret.sort_by_key(|f| index_map[&f.header]);
+        ret.into_iter()
     }
 
     pub fn extract<'a>(
         &self,
         bucket: &LoopBucket,
         recorder: EnvRecorder<'a, '_>,
-        context_kind: EnvContextKind,
         unrolled: &mut InstructionList,
     ) -> Result<(), BadInterp> {
         assert!(bucket.body.len() > 1);
-        let extra_arg_info = Self::compute_extra_args(&recorder, context_kind)?;
-        let name = self.build_new_body(
-            &recorder.get_current_scope_name(),
-            bucket,
-            extra_arg_info.bucket_to_args.clone(),
-            extra_arg_info.num_args,
-        );
+
+        // Check if an applicable function already exists, otherwise create a new extracted loop body function.
+        let (arg_info, mem_refs, extracted_name) = {
+            let (arg_info, mem_refs) = Self::compute_extra_args(&recorder)?;
+            let arg_info = Rc::new(arg_info);
+            let existing_fun = self.get_extracted_function_name(bucket.id, Rc::clone(&arg_info));
+            match existing_fun {
+                None => {
+                    // Create and store the new function
+                    let new_func = self.build_new_extracted_function(
+                        &recorder.get_current_scope_name(),
+                        bucket,
+                        &arg_info.loc_to_args,
+                        arg_info.num_args,
+                    );
+                    let new_name = new_func.header.clone();
+                    self.store_new_extracted_function(bucket.id, Rc::clone(&arg_info), new_func);
+
+                    (arg_info, mem_refs, new_name)
+                }
+                Some(name) => (arg_info, mem_refs, name.clone()),
+            }
+        };
+
+        // Store the parameter information for the function to GlobalPassData based on current Env
+        {
+            let mut gd = recorder.global_data.borrow_mut();
+            let extraction_data =
+                gd.extract_func_orig_loc.entry(extracted_name.clone()).or_default();
+            for iter_num in 0..recorder.get_iter() {
+                let iter_env = recorder.take_header_vars_for_iter(&iter_num);
+                let mapping = arg_info.get_reverse_passing_refs_for_itr(&mem_refs, iter_num);
+                if DEBUG_LOOP_UNROLL {
+                    println!(
+                        "[extract] storing orig loc data for: {}+{:?} -> {:?}",
+                        extracted_name, iter_env, mapping
+                    );
+                }
+                // NOTE: Encountering different iteration counts for the same loop will produce
+                //  different Env at loop header (i.e. `iter_env`) and the same Env at the loop
+                //  header will produce the same reverse mapping. Thus, no key conflicts.
+                checked_insert!(&mut *extraction_data, iter_env, mapping);
+            }
+        }
```

Comment:
Can we make this a method of `recorder` instead of making `global_data` public?

Reply:
I had done that before but I didn't like it. The function only accesses some data of `recorder` to store to the global map so I thought it was cleaner to move it out of `recorder` and just use the accessor functions. If you don't like the public field I could just make an accessor function for `recorder.global_data` but I thought this was cleaner than just adding a boilerplate function since the field is not mutable anyway so it cannot be accidentally misused by a future code change.

Reply:
I don't like getters for the sake of getters, so we can leave it as is then